server/service/src/rnr_form/finalise.rs

```diff
@@ -90,3 +102,89 @@ impl From<RepositoryError> for FinaliseRnRFormError {
+        requisition_number: next_number(
+            &ctx.connection,
+            &NumberRowType::RequestRequisition,
+            &ctx.store_id,
+        )?,
+        name_link_id: rnr_form.name_link_id,
+        store_id: rnr_form.store_id,
+        r#type: RequisitionType::Request,
+        status: RequisitionStatus::Sent,
+        created_datetime: Utc::now().naive_utc(),
+        sent_datetime: None,
+        finalised_datetime: None,
+        expected_delivery_date: None,
+        colour: None,
+        comment: Some("Automatically created from R&R Form".to_string()),
+        their_reference: None,
+        max_months_of_stock: 0.0,
+        min_months_of_stock: 0.0,
+        approval_status: None,
+        linked_requisition_id: None,
+        program_id: Some(rnr_form.program_id),
+        period_id: Some(rnr_form.period_id),
+        order_type: None, // Should we capture this in the RnR form?
+    };
+
+    let requisition_repo = RequisitionRowRepository::new(&ctx.connection);
+    requisition_repo.upsert_one(&requisition_row)?;
+
+    // Get R&R Form lines and create requisition lines
+
+    let rnr_form_line_repo = RnRFormLineRowRepository::new(&ctx.connection);
+    let requisition_line_repo = RequisitionLineRowRepository::new(&ctx.connection);
+    let item_repo = ItemRowRepository::new(&ctx.connection);
+
+    let rnr_form_lines = rnr_form_line_repo.find_many_by_rnr_form_id(&rnr_form.id)?;
+
+    // Find all the item names
+    let item_ids = rnr_form_lines
+        .iter()
+        .map(|rnr_form_line| rnr_form_line.item_id.clone())
+        .collect();
+
+    let items = item_repo.find_many_by_id(&item_ids)?;
+
+    let item_map = items
+        .into_iter()
+        .map(|item| (item.id.clone(), item))
+        .collect::<std::collections::HashMap<String, _>>();
+
+    // Loop through the rnr lines and create requisition lines
+
+    for rnr_form_line in rnr_form_lines {
+        let requisition_line = RequisitionLineRow {
+            id: uuid(),
+            requisition_id: requisition_row.id.clone(),
+            item_link_id: rnr_form_line.item_id.clone(),
+            item_name: item_map
+                .get(&rnr_form_line.item_id)
+                .map(|item| item.name.clone())
+                .unwrap_or_default(),
+            requested_quantity: rnr_form_line.requested_quantity,
+            suggested_quantity: rnr_form_line.requested_quantity, // TODO This is user enterable, should we also re-calc this?
+            supply_quantity: 0.0,
+            available_stock_on_hand: rnr_form_line.final_balance,
+            average_monthly_consumption: rnr_form_line.average_monthly_consumption,
+            snapshot_datetime: rnr_form.finalised_datetime,
+            approved_quantity: 0.0,
+            approval_comment: None,
+            comment: None,
+            // TODO add Cust_pre_stock_balance, Cust_stock_received, Cust_stock_ord, Cust_stock_adj (in mSupply but no OMS Yet)
```

Comment:
Should we bring these fields though open-mSupply? Maybe we can do it just at the translation layer in sync?

Reply:
I believe that we'll need them at some point. will need to discuss before we implement that though - I'm not sure why they haven't been done yet.